store/etcd/etcd.go

```diff
@@ -735,31 +766,71 @@ func (l *etcdLock) holdLock(key string, lockHeld chan struct{}, stopLocking <-ch
 	for {
 		select {
 		case <-update.C:
+			fmt.Printf("=====> %d: etcd.holdLock(): got update, setting %s to %s with prevIndex %d\n", getGID(), key, l.value, setOpts.PrevIndex)
 			setOpts.PrevIndex = l.last.Node.ModifiedIndex
 			l.last, err = l.client.Set(context.Background(), key, l.value, setOpts)
 			if err != nil {
+				fmt.Printf("=====> %d: etcd.holdLock(): got error %s on setting %s to %s\n", getGID(), err, key, l.value)
 				return
 			}
 
-		case <-stopLocking:
+		case msg := <-stopLocking:
+			fmt.Printf("=====> %d: etcd.holdLock(): got stopLocking: %p\n", getGID(), &msg)
 			return
 		}
 	}
 }
 
 // WaitLock simply waits for the key to be available for creation
 func (l *etcdLock) waitLock(key string, errorCh chan error, stopWatchCh chan bool, free chan<- bool) {
+
 	opts := &etcd.WatcherOptions{Recursive: false}
 	watcher := l.client.Watcher(key, opts)
-
+	if l.waitLockDelay == 0 {
+		l.waitLockDelay = 1 * time.Microsecond
+	}
+	retryLock := true
+	ctx, cancelFunc := context.WithTimeout(context.Background(), l.waitLockDelay)
+	defer cancelFunc()
 	for {
-		event, err := watcher.Next(context.Background())
+		fmt.Printf("=====> %d: In the waitLock loop, waiting for event\n", getGID())
+		event, err := watcher.Next(ctx)
 		if err != nil {
+			if err == context.DeadlineExceeded {
+				fmt.Printf("=====> %d: In the waitLock loop, got DeadlineExceeded\n", getGID())
+				// First, see if maybe the key is not there.
+				fmt.Printf("=====> %d: In the waitLock loop, checking for %s\n", getGID(), key)
+				kp, err2 := l.client.Get(ctx, key, nil)
+				if err2 != nil {
+					if err2 == store.ErrKeyNotFound {
+						fmt.Printf("=====> %d: In the waitLock loop, %s not found \n", getGID(), key)
+						retryLock = false
+					} else {
+						fmt.Printf("=====> %d: In the waitLock loop, checking on %s: %s (%T)", getGID(), key, err2, err2)
+					}
+				} else {
+					fmt.Printf("=====> %d: In the waitLock loop, got %v", getGID(), kp)
+				}
+				if !retryLock {
+					free <- true
+					l.waitLockDelay = 0
+					return
+				}
+
+				fmt.Printf("=====> %d: In the waitLock loop, got DeadlineExceeded, will retry for %s\n", getGID(), l.waitLockDelay)
+				l.waitLockDelay *= 2
```

Comment:
There is no upper bound on this. The exponential backoff can just keep growing. Maybe have an upper bound of a second, or so?

Should we make it error out if the total wait time exceeds 10 seconds, or something? With a log message? Because otherwise we're in this endless loop (if something goes wrong) and won't see anything in our logs (assuming a production cluster with cleaned up logs).

Reply:
Done